ndhc.8
.TH NDHC 8 2022-02-12 Linux "Linux Administrator's Manual"
.SH NAME
ndhc \- secure DHCPv4 client
.SH SYNOPSIS
.B ndhc
.RI [ OPTION ]...
.SH DESCRIPTION
The ndhc client negotiates a lease with the DHCPv4 server.  Once a lease is
obtained, it then defends the assigned IP address against hostile imposters and
requests a new lease if it detects that the interface has been connected to a
new network.
.SH OPTIONS
.TP
.BI \-c\  CONFIGFILE ,\ \-\-config= CONFIGFILE
Read configuration information from the specified file.  The format of
configuration options is a simple 'option = value' for each line.  The
names of the options are exactly the same as for command line options.
.TP
.BI \-I\  CLIENTID ,\ \-\-clientid= CLIENTID
Specifies the client identifier that will be sent to the remote server.  This
can be any (reasonably sized, <64byte or so) text string, or an ethernet
MAC address in a form similar to 'aa:bb:cc:dd:ee:ff'.  ndhc is smart enough
to recognize MAC addresses.  ISP DHCP servers commonly check the value of this
field before providing a lease.  The default value is the MAC address of
the network interface to which ndhc is bound.
.TP
.BI \-h\  HOSTNAME ,\ \-\-hostname= HOSTNAME
Send the specified client hostname to the remote DHCP server.  This option
should not be necessary in most instances, but may perhaps be useful for odd
DHCP servers that perform some kind of authentication against the hostname
option field.  The default is to send no hostname option at all.
.TP
.BI \-v\  VENDORID ,\ \-\-vendorid= VENDORID
Send the specified vendor identification string to the remote DHCP server.
This option should not be necessary in most instances, but may perhaps be
useful for odd DHCP servers that perform some kind of authentication against
the vendor id option field.  The default is to send the string 'ndhc'.
.TP
.BI \-s\  STATEDIR ,\ \-\-state\-dir= STATEDIR
Specifies the directory where the DHCP state associated with the given
interface will be stored.  Such state will include the leased IP, the
IAID, and the DUID.  The file representing the leased IP can be quite
useful for reacting to changes in IP address -- one can listen for changes
to it using fanotify() or inotify() on Linux.
.TP
.BI \-i\  INTERFACE ,\ \-\-interface= INTERFACE
Act as a DHCP client for the specified interface.  A single ndhc daemon can
only act as a DHCP client for a single interface.  Specify the interface it
should use by name.  The default is to listen on 'eth0'.
.TP
.BI \-n ,\  \-\-now
Exit with failure if a lease cannot be obtained.  Useful for some init scripts.
.TP
.BI \-r\  IP ,\ \-\-request= IP
Request the specified IP address from the remote DHCP server.  The DHCP server
has no obligation to provide us with this IP, but it may acquiesce to the
request if it would not conflict with another host.
.TP
.BI \-u\  USER ,\ \-\-user= USER
This option specifies the user name or user id that ndhc will change to after
startup.  ndhc will also change its group to match the default group of this
user.
.TP
.BI \-U\  USER ,\ \-\-ifch\-user= USER
This option specifies the user name or user id that ndhc-ifch will change to
after startup.  ndhc-ifch will also change its group to match the default group
of this user.
.TP
.BI \-C\  CHROOTDIR ,\ \-\-chroot= CHROOTDIR
This option specifies the directory to which ndhc should confine itself via
chroot() after startup.  This directory should have access to dev/urandom and
dev/null.  For logging to work, a dev/log socket or device should also exist.
.TP
.BI \-d ,\  \-\-relentless\-defense
If specified, ndhc will never back down in defending the IP address that it
has been assigned by the remote DHCP server.  This behavior should not be
specified for average machines, but is useful for servers or routers where
the IP address of the machine must remain fixed for proper operation.
.TP
.BI \-R\  RESOLVCONF ,\ \-\-resolv\-conf= RESOLVCONF
Specifies the path to the system resolv.conf.  This file will typically be in
/etc/resolv.conf.  If this option is specified, ndhc will update the contents
of this file to match the DNS servers specified by the remote DHCP server.  If
this option is not specified, ifchd will never change the system DNS resolution
configuration.
.TP
.BI \-H ,\  \-\-dhcp\-set\-hostname
If specified, ndhc will update the system host name in response to any
hostname option field provided by a remote DHCP server on the request of
a ndhc client.  If this option is not specified, ndhc will never change
the system hostname.
.TP
.BI \-w\  TIMEMS ,\  \-\-arp\-probe\-wait= TIMEMS
Adjusts the time that we wait for an ARP response when checking to see if
our lease assignment is already taken by an existing host.  Default is
1000ms.
.TP
.BI \-W\  NUMPROBES ,\  \-\-arp\-probe\-num= NUMPROBES
Adjusts the number of ARP packets that we send when probing for collisions
with an existing host that is using our assigned IP.  Once we have sent
the specified number of probe packets with no response, ndhc is willing
to believe that there is no colliding host.  Default number is 3 probes.
.TP
.BI \-m\  TIMEMS ,\  \-\-arp\-probe\-min= TIMEMS
Adjusts the minimum time that we wait between sending probe packets.  The
default is 1000ms.  The precise inter-probe wait time is randomized.
.TP
.BI \-M\  TIMEMS ,\  \-\-arp\-probe\-max= TIMEMS
Adjusts the maximum time that we wait between sending probe packets.  The
default is 2000ms.  The precise inter-probe wait time is randomized.
.TP
.BI \-t\  GWMETRIC ,\  \-\-gw\-metric= GWMETRIC
Specifies the routing metric for the default gateway entry.  Defaults to
0 if not specified.  Higher values will de-prioritize the route entry.
.TP
.BI \-K\  RFKILLIDX ,\  \-\-rfkill\-idx= RFKILLIDX
If set, specifies the rfkill device index that corresponds to this interface.
ndhc will then listen for matching radio frequency kill switch events
and will bring the interface up and down in reaction to the events.
The rfkill devices can be found in /sys/class/rfkill/rfkill<RFKILLIDX>.
It may be useful to check the contents of the 'name' file within this
directory to determine the correct device index.  In any event, if
an rfkill-idx parameter is specified, ndhc will print messages for any
rfkill events that it sees, so it should not be too difficult to locate
the proper rfkill device by checking the logs after hitting the switch.
.TP
.BI \-N\  NOTIFY_FDNUM ,\  \-\-s6\-notify= NOTIFY_FDNUM
If set, specifies the file descriptor number that will have a '\n' written to
and closed when the first DHCP lease is bound.  This option should be used when
ndhc is run under a s6 supervisor that implements service startup
notifications.  Note that if this option is set and no notification file
descriptor is passed to the ndhc process it is very likely that ndhc will
fail to run properly in cryptic and hard to debug ways.
.TP
.BI \-X\  SCRIPTFILE ,\  \-\-script\-file= SCRIPTFILE
If set, ndhc will spawn a subprocess that has the exclusive job of executing
the specified executable file immediately after a new lease is acquired.  This
script file will run as root and will not be chrooted.  It will be provided a
sanitized environment that has no inputs from the dhcp state.  If this
parameter is not provided, then the ndhc-scriptd subprocess will not exist.
This facility is intended to be used for updating firewall/nat rules or similar
tasks.
.TP
.BI \-v ,\  \-\-version
Display the ndhc version number.
.SH SIGNALS
It is not necessary to sleep between sending signals, as signals received are
processed sequentially in the order they are received.
.B ndhc
responds to the following signals:
.TP
.B SIGUSR1
This signal causes
.B ndhc
to renew the current lease or, if it does not have one, obtain a
new lease.
.TP
.B SIGUSR2
This signal causes
.B ndhc
to release the current lease and go to sleep until it receives a SIGUSR1.
.SH NOTES
ndhc will seed its random number generator (used for generating xids)
by reading /dev/urandom. If you have a lot of embedded systems on the same
network, with no entropy, you can either seed /dev/urandom by a method of
your own, or doing the following on startup:

ifconfig eth0 > /dev/urandom

in order to seed /dev/urandom with some data (mac address) unique to your
system. If reading /dev/urandom fails, ndhc will fall back to seeding with
time(0).



ChangeLog
20040906
	accept_conns() trivial correctness fix: EINTR return from accept should
	force continue rather than break in handling loop.

	fail_on_fdne() is now file_exists() with a slightly different API.

20040818
	Use inet_aton rather than custom function; daemon.[ch] removed.

20040817
	ifchd no longer logs a "FATAL - select returned an error!" when it
	is -KILL'ed.

20040626
	pending_exit is now static volatile sig_atomic_t rather than
	static volatile int for pedantic correctness issues.

20040614
	Added --interface (-i) option that allows one to restrict the
		interfaces that ifchd clients are allowed to modify.  I
		reccomend that this flag be used to further mitigate the possible
		effects of a compromised client.  By default, all interfaces
		may be modified by clients.
	Minor cleanups.
	Clear corresponding namesvrs and domains on socket disconnection.

20040613
	Factor out Linux-specific code into linux.c.
	Nameservers and search domains now tracked per-connection to
		prevent races where a client could force a writeout of data
		provided by another client.
	Interface name now cleared on connection close.
	Make all headers idempotent.
	Minor improvements to error messages.

20040612
	Removed iffd[] array; this change makes the state machine action
	functions (perform_*) depend on less external state.

	HOSTNAME command now supported; it is not enabled by default.
	If you wish for remote daemons to be able to change the hostname
	of the local machine, use the --hostname (-o) option.

20040610
	Added MTU support.

20040609
	Robustify so that suicide() isn't called at every possible failure.
	Harmless failures will now simply print an error to the log.  Risky
	or severe errors still suicide().

20040608

	Added support for resolv.conf, namely "nameserver" and "search" entries.
	Make strlist more robust and paranoid.

20040607

	Improved accept() code to be much more robust; it can now properly error
	recover.  I hope this will fix the elusive random-exit problem that
	plagues machines running the 2.6 kernel.

	Refactored dispatch_work() to be cleaner.



ndhc.c
// Copyright 2004-2022 Nicholas J. Kain <njkain at gmail dot com>
// SPDX-License-Identifier: MIT
#include <stdio.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/file.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h>
#include <time.h>
#include <string.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <poll.h>
#include <sys/prctl.h>
#include <net/if.h>
#include <errno.h>
#include <pwd.h>
#include <grp.h>
#include <limits.h>
#include "nk/log.h"
#include "nk/privs.h"
#include "nk/io.h"

#include "ndhc.h"
#include "ndhc-defines.h"
#include "cfg.h"
#include "state.h"
#include "options.h"
#include "dhcp.h"
#include "sys.h"
#include "ifchange.h"
#include "arp.h"
#include "nl.h"
#include "netlink.h"
#include "leasefile.h"
#include "ifset.h"
#include "ifchd.h"
#include "duiaid.h"
#include "sockd.h"
#include "rfkill.h"
#include "scriptd.h"

struct client_state_t cs = {
    .program_init = true,
    .listenFd = -1,
    .arpFd = -1,
    .nlFd = -1,
    .nlPortId = 0,
    .rfkillFd = -1,
    .dhcp_wake_ts = -1,
    .routerArp = "\0\0\0\0\0\0",
    .serverArp = "\0\0\0\0\0\0",
};

struct client_config_t client_config = {
    .interface = "eth0",
    .arp = "\0\0\0\0\0\0",
    .s6_notify_fd = 3,
    .clientid_len = 0,
    .metric = 0,
};

static volatile sig_atomic_t l_signal_exit;
static volatile sig_atomic_t l_signal_renew;
static volatile sig_atomic_t l_signal_release;
// Intended to be called in a loop until SIGNAL_NONE is returned.
int signals_flagged(void)
{
    if (l_signal_exit) {
        l_signal_exit = 0;
        return SIGNAL_EXIT;
    }
    if (l_signal_renew) {
        l_signal_renew = 0;
        return SIGNAL_RENEW;
    }
    if (l_signal_release) {
        l_signal_release = 0;
        return SIGNAL_RELEASE;
    }
    return SIGNAL_NONE;
}

bool carrier_isup(void) { return cs.carrier_up; }

void set_client_addr(const char *v) { cs.clientAddr = inet_addr(v); }

void print_version(void)
{
    printf("ndhc %s, dhcp client.\n", NDHC_VERSION);
    printf("Copyright 2004-2022 Nicholas J. Kain\n\n"
"Permission is hereby granted, free of charge, to any person obtaining\n"
"a copy of this software and associated documentation files (the\n"
"\"Software\"), to deal in the Software without restriction, including\n"
"without limitation the rights to use, copy, modify, merge, publish,\n"
"distribute, sublicense, and/or sell copies of the Software, and to\n"
"permit persons to whom the Software is furnished to do so, subject to\n"
"the following conditions:\n\n"
"The above copyright notice and this permission notice shall be\n"
"included in all copies or substantial portions of the Software.\n\n"
"THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND,\n"
"EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF\n"
"MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND\n"
"NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE\n"
"LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION\n"
"OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION\n"
"WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n"
           );
    exit(EXIT_SUCCESS);
}

void show_usage(void)
{
    printf(
"ndhc " NDHC_VERSION ", dhcp client.\n"
"Copyright 2004-2022 Nicholas J. Kain\n"
"Usage: ndhc [OPTIONS]\n\n"
"  -c, --config=FILE               Path to ndhc configuration file\n"
"  -I, --clientid=CLIENTID         Client identifier\n"
"  -h, --hostname=HOSTNAME         Client hostname\n"
"  -V, --vendorid=VENDORID         Client vendor identification string\n"
"  -i, --interface=INTERFACE       Interface to use (default: eth0)\n"
"  -n, --now                       Exit with failure if lease cannot be\n"
"                                  immediately negotiated.\n"
"  -r, --request=IP                IP address to request (default: none)\n"
"  -u, --user=USER                 ndhc runs as this user\n"
"  -U, --ifch-user=USER            ndhc-ifch runs as this user\n"
"  -D, --sockd-user=USER           ndhc-sockd runs as this user\n"
"  -C, --chroot=DIR                Chroot to this directory\n"
"  -s, --state-dir=DIR             State storage dir (default: /etc/ndhc)\n"
"  -d, --relentless-defense        Never back off in defending IP against\n"
"                                  conflicting hosts (servers only)\n"
"  -w, --arp-probe-wait            Time to delay before first ARP probe\n"
"  -W, --arp-probe-num             Number of ARP probes before lease is ok\n"
"  -m, --arp-probe-min             Min ms to wait for ARP response\n"
"  -M, --arp-probe-max             Max ms to wait for ARP response\n"
"  -t, --gw-metric                 Route metric for default gw (default: 0)\n"
"  -R, --resolve-conf=FILE         Path to resolv.conf or equivalent\n"
"  -H, --dhcp-set-hostname         Allow DHCP to set machine hostname\n"
"  -v, --version                   Display version\n"
           );
    exit(EXIT_SUCCESS);
}

static void signal_handler(int signo)
{
    int serrno = errno;
    switch (signo) {
    case SIGCHLD: {
        static const char errstr[] = "ndhc-master: Subprocess terminated unexpectedly. Exiting.\n";
        safe_write(STDOUT_FILENO, errstr, sizeof errstr - 1);
        _exit(EXIT_FAILURE);
    }
    case SIGINT:
    case SIGTERM: l_signal_exit = 1; break;
    case SIGUSR1: l_signal_renew = 1; break;
    case SIGUSR2: l_signal_release = 1; break;
    default: break;
    }
    errno = serrno;
}

void signal_exit(int status)
{
    log_line("Received terminal signal. Exiting.\n");
    exit(status);
}

static void setup_signals_ndhc(void)
{
    static const int ss[] = {
        SIGCHLD, SIGINT, SIGTERM, SIGUSR1, SIGUSR2, SIGKILL
    };
    sigset_t mask;

    if (sigprocmask(0, 0, &mask) < 0)
        suicide("sigprocmask failed\n");
    for (int i = 0; ss[i] != SIGKILL; ++i)
        if (sigdelset(&mask, ss[i]))
            suicide("sigdelset failed\n");
    if (sigaddset(&mask, SIGPIPE))
        suicide("sigaddset failed\n");
    if (sigprocmask(SIG_SETMASK, &mask, (sigset_t *)0) < 0)
        suicide("sigprocmask failed\n");

    struct sigaction sa = {
        .sa_handler = signal_handler,
        .sa_flags = SA_RESTART,
    };
    if (sigemptyset(&sa.sa_mask))
        suicide("sigemptyset failed\n");
    for (int i = 0; ss[i] != SIGKILL; ++i)
        if (sigaction(ss[i], &sa, NULL))
            suicide("sigaction failed\n");
}

static void fail_if_state_dir_dne(void)
{
    if (strlen(state_dir) == 0)
        suicide("state_dir path is empty; it must be specified\n");
    struct stat st;
    if (stat(state_dir, &st) < 0)
        suicide("failed to stat state_dir path '%s': %s\n",
                state_dir, strerror(errno));
    if (!S_ISDIR(st.st_mode))
        suicide("state_dir path '%s' does not specify a directory\n", state_dir);
}

static void do_ndhc_work(void)
{
    static bool rfkill_set; // Is the rfkill switch set?
    static bool rfkill_nl_carrier_wentup; // iface carrier changed to up during rfkill
    struct dhcpmsg dhcp_packet;
    long long nowts;
    int timeout = 0;
    bool had_event;

    setup_signals_ndhc();
    start_dhcp_listen(&cs);

    struct pollfd pfds[] = {
        [0] = { .fd = cs.nlFd,          .events = POLLIN|POLLHUP|POLLERR|POLLRDHUP },
        [1] = { .fd = ifchStream[0],    .events = POLLHUP|POLLERR|POLLRDHUP },
        [2] = { .fd = sockdStream[0],   .events = POLLHUP|POLLERR|POLLRDHUP },
        [6] = { .fd = scriptdStream[0], .events = POLLHUP|POLLERR|POLLRDHUP },
        [3] = { .fd = cs.rfkillFd,      .events = POLLIN|POLLHUP|POLLERR|POLLRDHUP },
        // These can change on the fly.
        [4] = { .events = POLLIN|POLLHUP|POLLERR|POLLRDHUP },
        [5] = { .events = POLLIN|POLLHUP|POLLERR|POLLRDHUP },
    };
    for (;;) {
        pfds[4].fd = cs.arpFd;
        pfds[5].fd = cs.listenFd;
        had_event = false;
        if (poll(pfds, 7, timeout) < 0) {
            if (errno != EINTR) suicide("poll failed\n");
        }

        bool sev_dhcp = false;
        uint32_t dhcp_srcaddr = 0;
        uint8_t dhcp_msgtype = 0;
        bool sev_arp = false;
        int sev_nl = IFS_NONE;
        int sev_rfk = RFK_NONE;
        bool force_fingerprint = false;
        if (pfds[0].revents & POLLIN) {
            had_event = true;
            sev_nl = nl_event_get(&cs);
            if (!cs.carrier_up)
                cs.carrier_up = (sev_nl == IFS_UP);
        }
        if (pfds[0].revents & (POLLHUP|POLLERR|POLLRDHUP)) {
            suicide("nlfd closed unexpectedly\n");
        }
        if (pfds[1].revents & (POLLHUP|POLLERR|POLLRDHUP)) {
            exit(EXIT_FAILURE);
        }
        if (pfds[2].revents & (POLLHUP|POLLERR|POLLRDHUP)) {
            exit(EXIT_FAILURE);
        }
        if (pfds[6].revents & (POLLHUP|POLLERR|POLLRDHUP)) {
            exit(EXIT_FAILURE);
        }
        if (pfds[3].revents & POLLIN) {
            had_event = true;
            sev_rfk = rfkill_get(&cs, 1, client_config.rfkillIdx);
        }
        if (pfds[3].revents & (POLLHUP|POLLERR|POLLRDHUP)) {
            suicide("rfkillfd closed unexpectedly\n");
        }
        if (pfds[4].revents & POLLIN) {
            had_event = true;
            // Make sure the fd is still the same.
            if (pfds[4].fd == cs.arpFd)
                sev_arp = arp_packet_get(&cs);
        }
        if (pfds[4].revents & (POLLHUP|POLLERR|POLLRDHUP)) {
            suicide("arpfd closed unexpectedly\n");
        }
        if (pfds[5].revents & POLLIN) {
            had_event = true;
            // Make sure the fd is still the same.
            if (pfds[5].fd == cs.listenFd)
                sev_dhcp = dhcp_packet_get(&cs, &dhcp_packet, &dhcp_msgtype,
                                           &dhcp_srcaddr);
        }
        if (pfds[5].revents & (POLLHUP|POLLERR|POLLRDHUP)) {
            suicide("listenfd closed unexpectedly\n");
        }

        if (sev_rfk == RFK_ENABLED) {
            rfkill_set = 1;
            rfkill_nl_carrier_wentup = false;
            log_line("rfkill: radio now blocked\n");
        } else if (sev_rfk == RFK_DISABLED) {
            rfkill_set = 0;
            log_line("rfkill: radio now unblocked\n");
            cs.carrier_up = ifchange_carrier_isup();
            if (rfkill_nl_carrier_wentup && carrier_isup()) {
                // We might have changed networks while the radio was down.
                force_fingerprint = true;
            }
        }

        if (sev_nl != IFS_NONE && nl_event_carrier_wentup(sev_nl)) {
            if (!rfkill_set)
                force_fingerprint = true;
            else
                rfkill_nl_carrier_wentup = true;
        }

        if (rfkill_set || !carrier_isup()) {
            // We can't do anything while the iface is disabled, anyway.
            // Suspend might cause link state change notifications to be
            // missed, so we use a non-infinite timeout.
            timeout = 2000 + (int)(nk_random_u32(&cs.rnd_state) % 3000);
            continue;
        }

        // These two can change on the fly; make sure the event is current.
        if (pfds[4].fd != cs.arpFd) sev_arp = false;
        if (pfds[5].fd != cs.listenFd) sev_dhcp = false;

        nowts = curms();
        long long arp_wake_ts = arp_get_wake_ts();
        int dhcp_ok = dhcp_handle(&cs, nowts, sev_dhcp, &dhcp_packet,
                                  dhcp_msgtype, dhcp_srcaddr,
                                  sev_arp, force_fingerprint,
                                  cs.dhcp_wake_ts <= nowts,
                                  arp_wake_ts <= nowts);

        if (dhcp_ok == COR_ERROR) {
            timeout = 2000 + (int)(nk_random_u32(&cs.rnd_state) % 3000);
            continue;
        }

        int prev_timeout = timeout;
        long long tt;

        arp_wake_ts = arp_get_wake_ts();
        if (arp_wake_ts < 0 && cs.dhcp_wake_ts < 0) {
            timeout = -1;
            continue;
        } else if (arp_wake_ts < 0) {
            tt = cs.dhcp_wake_ts - nowts;
        } else if (cs.dhcp_wake_ts < 0) {
            tt = arp_wake_ts - nowts;
        } else {
            tt = (arp_wake_ts < cs.dhcp_wake_ts ?
                  arp_wake_ts : cs.dhcp_wake_ts) - nowts;
        }
        if (tt > INT_MAX) tt = INT_MAX;
        if (tt < INT_MIN) tt = INT_MIN;
        timeout = tt;
        if (timeout < 0)
            timeout = 0;

        // Failsafe to prevent busy-spin.
        if (timeout == 0 && prev_timeout == 0 && !had_event)
            timeout = 10000;
    }
}

char state_dir[PATH_MAX] = "/etc/ndhc";
char chroot_dir[PATH_MAX] = "";
char resolv_conf_d[PATH_MAX] = "";
char script_file[PATH_MAX] = "";
uid_t ndhc_uid = 0;
gid_t ndhc_gid = 0;
int ifchSock[2];
int ifchStream[2];
int sockdSock[2];
int sockdStream[2];
int scriptdSock[2];
int scriptdStream[2] = { -1, -1 };

static void create_ifch_ipc_sockets(void) {
    if (socketpair(AF_UNIX, SOCK_DGRAM, 0, ifchSock) < 0)
        suicide("FATAL - can't create ndhc/ifch socket: %s\n", strerror(errno));
    if (socketpair(AF_UNIX, SOCK_STREAM, 0, ifchStream) < 0)
        suicide("FATAL - can't create ndhc/ifch socket: %s\n", strerror(errno));
}

static void create_sockd_ipc_sockets(void) {
    if (socketpair(AF_UNIX, SOCK_DGRAM, 0, sockdSock) < 0)
        suicide("FATAL - can't create ndhc/sockd socket: %s\n", strerror(errno));
    if (socketpair(AF_UNIX, SOCK_STREAM, 0, sockdStream) < 0)
        suicide("FATAL - can't create ndhc/sockd socket: %s\n", strerror(errno));
}

static void create_scriptd_ipc_sockets(void) {
    if (socketpair(AF_UNIX, SOCK_DGRAM, 0, scriptdSock) < 0)
        suicide("FATAL - can't create ndhc/scriptd socket: %s\n", strerror(errno));
    if (socketpair(AF_UNIX, SOCK_STREAM, 0, scriptdStream) < 0)
        suicide("FATAL - can't create ndhc/scriptd socket: %s\n", strerror(errno));
}

static void spawn_ifch(void)
{
    create_ifch_ipc_sockets();
    pid_t ifch_pid = fork();
    if (ifch_pid == 0) {
        close(ifchSock[0]);
        close(ifchStream[0]);
        // Don't share the RNG state with the master process.
        nk_random_init(&cs.rnd_state);
        ifch_main();
    } else if (ifch_pid > 0) {
        close(ifchSock[1]);
        close(ifchStream[1]);
    } else
        suicide("failed to fork ndhc-ifch: %s\n", strerror(errno));
}

static void spawn_sockd(void)
{
    create_sockd_ipc_sockets();
    pid_t sockd_pid = fork();
    if (sockd_pid == 0) {
        close(sockdSock[0]);
        close(sockdStream[0]);
        // Don't share the RNG state with the master process.
        nk_random_init(&cs.rnd_state);
        sockd_main();
    } else if (sockd_pid > 0) {
        close(sockdSock[1]);
        close(sockdStream[1]);
    } else
        suicide("failed to fork ndhc-sockd: %s\n", strerror(errno));
}

static void spawn_scriptd(void)
{
    valid_script_file = access(script_file, R_OK | X_OK) == 0;
    if (!valid_script_file) return;

    log_line("Found script file: '%s'\n", script_file);

    create_scriptd_ipc_sockets();
    pid_t scriptd_pid = fork();
    if (scriptd_pid == 0) {
        close(scriptdSock[0]);
        close(scriptdStream[0]);
        // Don't share the RNG state with the master process.
        nk_random_init(&cs.rnd_state);
        scriptd_main();
    } else if (scriptd_pid > 0) {
        close(scriptdSock[1]);
        close(scriptdStream[1]);
    } else
        suicide("failed to fork ndhc-scriptd: %s\n", strerror(errno));
}

static void ndhc_main(void) {
    prctl(PR_SET_NAME, "ndhc: master");
    log_line("ndhc client " NDHC_VERSION " started on interface [%s].\n",
             client_config.interface);

    if ((cs.nlFd = nl_open(NETLINK_ROUTE, RTMGRP_LINK, &cs.nlPortId)) < 0)
        suicide("%s: failed to open netlink socket\n", __func__);

    cs.rfkillFd = rfkill_open(&client_config.enable_rfkill);

    open_leasefile();

    nk_set_chroot(chroot_dir);
    memset(chroot_dir, '\0', sizeof chroot_dir);
    nk_set_uidgid(ndhc_uid, ndhc_gid, (const unsigned char *)0, 0);

    cs.carrier_up = ifchange_carrier_isup();
    if (!carrier_isup()) {
        if (ifchange_deconfig(&cs) < 0)
            suicide("%s: can't deconfigure interface settings\n", __func__);
    }

    do_ndhc_work();
}

static void wait_for_rfkill()
{
    cs.rfkillFd = rfkill_open(&client_config.enable_rfkill);
    if (cs.rfkillFd < 0)
        suicide("can't wait for rfkill to end if /dev/rfkill can't be opened\n");

    struct pollfd pfds[1] = {0};
    pfds[0].events = POLLIN|POLLHUP|POLLERR|POLLRDHUP;
    for (;;) {
        pfds[0].fd = cs.rfkillFd;
        if (poll(pfds, 1, -1) < 0) {
            if (errno != EINTR) suicide("poll failed\n");
        }
        if (pfds[0].revents & POLLIN) {
            if (rfkill_get(&cs, 0, 0) == RFK_DISABLED) {
                switch (perform_ifup()) {
                case 1:
                case 0: goto rfkill_gone;
                case -3:
                    log_line("rfkill: radio immediately blocked again; spurious?\n");
                    break;
                default: suicide("failed to set the interface to up state\n");
                }
            }
        }
        if (pfds[0].revents & (POLLHUP|POLLERR|POLLRDHUP)) {
            suicide("rfkillFd closed unexpectedly\n");
        }
    }
rfkill_gone:
    // We always close because ifchd and sockd shouldn't keep
    // an rfkill fd open.
    close(cs.rfkillFd);
    cs.rfkillFd = -1;
}

int main(int argc, char *argv[])
{
    parse_cmdline(argc, argv);

    nk_random_init(&cs.rnd_state);
    cs.xid = nk_random_u32(&cs.rnd_state);

    if (getuid())
        suicide("I need to be started as root.\n");
    if (!strncmp(chroot_dir, "", sizeof chroot_dir))
        suicide("No chroot path is specified.  Refusing to run.\n");
    fail_if_state_dir_dne();

    if (nl_getifdata() < 0)
        suicide("failed to get interface MAC or index\n");

    get_clientid(&client_config);

    switch (perform_ifup()) {
    case 1: case 0: break;
    case -3: wait_for_rfkill(); break;
    default: suicide("failed to set the interface to up state\n");
    }

    if (setpgid(0, 0) < 0) {
        // EPERM is returned if we are already a process group leader.
        if (errno != EPERM)
            suicide("setpgid failed: %s\n", strerror(errno));
    }

    spawn_ifch();
    spawn_sockd();
    spawn_scriptd();
    ndhc_main();
    return 0;
}

